File: products/yappc/frontend/libs/yappc-ai/src/agents/AgentOrchestrator.ts
/**
 * Agent Orchestrator
 *
 * Coordinates multiple AI agents for complex, multi-step workflows.
 * Handles dependency resolution, parallel execution, and error propagation.
 *
 * @module ai/agents/AgentOrchestrator
 * @doc.type class
 * @doc.purpose Multi-agent workflow coordination
 * @doc.layer product
 * @doc.pattern Mediator
 */

import { v4 as uuidv4 } from 'uuid';
import type {
    AgentContext,
    AgentName,
    AgentResult,
    IAIAgent,
} from './types';

/**
 * Workflow step definition
 */
export interface WorkflowStep {
    id: string;
    agentName: AgentName;
    input: Record<string, unknown> | string; // String for references like '@predict'
    dependsOn?: string[];
    optional?: boolean;
    timeout?: number;
}

/**
 * Agent workflow definition
 */
export interface AgentWorkflow {
    name: string;
    description?: string;
    steps: WorkflowStep[];
    stopOnError?: boolean;
    timeout?: number;
}

/**
 * Workflow execution result
 */
export interface WorkflowResult {
    success: boolean;
    workflowName: string;
    results: Map<string, AgentResult<unknown>>;
    metrics: WorkflowMetrics;
    errors?: WorkflowError[];
}

/**
 * Workflow metrics
 */
export interface WorkflowMetrics {
    totalDurationMs: number;
    stepDurations: Record<string, number>;
    agentsExecuted: number;
    successCount: number;
    failureCount: number;
}

/**
 * Workflow error
 */
export interface WorkflowError {
    stepId: string;
    agentName: AgentName;
    error: string;
    recoverable: boolean;
}

/**
 * Execution stage (steps that can run in parallel)
 */
interface ExecutionStage {
    steps: WorkflowStep[];
}

/**
 * Agent Orchestrator for coordinating multi-agent workflows
 */
export class AgentOrchestrator {
    private agents: Map<AgentName, IAIAgent<unknown, unknown>>;

    constructor() {
        this.agents = new Map();
    }

    /**
     * Register an agent with the orchestrator
     */
    registerAgent<TInput, TOutput>(agent: IAIAgent<TInput, TOutput>): void {
        this.agents.set(agent.name, agent as IAIAgent<unknown, unknown>);
    }

    /**
     * Get a registered agent
     */
    getAgent<TInput, TOutput>(name: AgentName): IAIAgent<TInput, TOutput> | undefined {
        return this.agents.get(name) as IAIAgent<TInput, TOutput> | undefined;
    }

    /**
     * List all registered agents
     */
    listAgents(): AgentName[] {
        return Array.from(this.agents.keys());
    }

    /**
     * Execute a multi-agent workflow
     */
    async executeWorkflow(
        workflow: AgentWorkflow,
        context: AgentContext
    ): Promise<WorkflowResult> {
        const startTime = Date.now();
        const results = new Map<string, AgentResult<unknown>>();
        const stepDurations: Record<string, number> = {};
        const errors: WorkflowError[] = [];

        // Build execution plan (resolve dependencies into stages)
        const executionPlan = this.buildExecutionPlan(workflow);

        // Execute stages sequentially, steps within stage in parallel
        for (const stage of executionPlan) {
            const stageResults = await Promise.all(
                stage.steps.map(async (step) => {
                    const stepStart = Date.now();
                    try {
                        const result = await this.executeStep(step, results, context);
                        stepDurations[step.id] = Date.now() - stepStart;
                        results.set(step.id, result);
                        return { step, result, success: result.success };
                    } catch (error) {
                        stepDurations[step.id] = Date.now() - stepStart;
                        const errorResult: AgentResult<unknown> = {
                            success: false,
                            error: error instanceof Error ? error : new Error(String(error)),
                            metrics: { latencyMs: stepDurations[step.id], modelVersion: 'unknown' },
                            trace: {
                                agentName: step.agentName,
                                requestId: context.requestId,
                                timestamp: new Date(),
                                metadata: { stepId: step.id, error: true },
                            },
                        };
                        results.set(step.id, errorResult);
                        return { step, result: errorResult, success: false };
                    }
                })
            );

            // Check for failures
            const failures = stageResults.filter((r) => !r.success && !r.step.optional);
            for (const failure of failures) {
                errors.push({
                    stepId: failure.step.id,
                    agentName: failure.step.agentName,
                    error: failure.result.error?.message || 'Unknown error',
                    recoverable: false,
                });
            }

            // Stop if configured and failures occurred
            if (workflow.stopOnError && failures.length > 0) {
                break;
            }
        }

        const successCount = Array.from(results.values()).filter((r) => r.success).length;
        const failureCount = results.size - successCount;

        return {
            success: failureCount === 0,
            workflowName: workflow.name,
            results,
            metrics: {
                totalDurationMs: Date.now() - startTime,
                stepDurations,
                agentsExecuted: results.size,
                successCount,
                failureCount,
            },
            errors: errors.length > 0 ? errors : undefined,
        };
    }

    /**
     * Build execution plan from workflow
     */
    private buildExecutionPlan(workflow: AgentWorkflow): ExecutionStage[] {
        const stages: ExecutionStage[] = [];
        const completed = new Set<string>();
        const remaining = new Set(workflow.steps.map((s) => s.id));

        while (remaining.size > 0) {
            // Find steps whose dependencies are all completed
            const readySteps = workflow.steps.filter(
                (step) =>
                    remaining.has(step.id) &&
                    (step.dependsOn || []).every((dep) => completed.has(dep))
            );

            if (readySteps.length === 0 && remaining.size > 0) {
                throw new Error('Circular dependency detected in workflow');
            }

            stages.push({ steps: readySteps });

            for (const step of readySteps) {
                completed.add(step.id);
                remaining.delete(step.id);
            }
        }

        return stages;
    }

    /**
     * Execute a single workflow step
     */
    private async executeStep(
        step: WorkflowStep,
        previousResults: Map<string, AgentResult<unknown>>,
        context: AgentContext
    ): Promise<AgentResult<unknown>> {
        const agent = this.agents.get(step.agentName);
        if (!agent) {
            throw new Error(`Agent not found: ${step.agentName}`);
        }

        // Build input from step config and previous results
        const input = this.buildStepInput(step, previousResults);

        // Apply step-specific timeout
        const stepContext: AgentContext = {
            ...context,
            timeout: step.timeout || context.timeout,
            requestId: `${context.requestId}-${step.id}`,
        };

        return agent.execute(input, stepContext);
    }

    /**
     * Build step input by resolving references to previous results
     */
    private buildStepInput(
        step: WorkflowStep,
        previousResults: Map<string, AgentResult<unknown>>
    ): unknown {
        if (typeof step.input === 'string' && step.input.startsWith('@')) {
            // Reference to previous step result
            const refId = step.input.slice(1);
            const result = previousResults.get(refId);
            if (!result || !result.success) {
                throw new Error(`Referenced step '${refId}' not found or failed`);
            }
            return result.data;
        }

        if (typeof step.input !== 'object' || step.input === null) {
            return step.input;
        }

        // Deep resolve references in object
        return this.resolveReferences(step.input, previousResults);
    }

    /**
     * Resolve references in an object
     */
    private resolveReferences(
        obj: Record<string, unknown>,
        previousResults: Map<string, AgentResult<unknown>>
    ): Record<string, unknown> {
        const result: Record<string, unknown> = {};

        for (const [key, value] of Object.entries(obj)) {
            if (typeof value === 'string' && value.startsWith('@')) {
                const refId = value.slice(1);
                const refResult = previousResults.get(refId);
                if (refResult?.success) {
                    result[key] = refResult.data;
                } else {
                    result[key] = undefined;
                }
            } else if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
                result[key] = this.resolveReferences(
                    value as Record<string, unknown>,
                    previousResults
                );
            } else {
                result[key] = value;
            }
        }

        return result;
    }

    /**
     * Execute a predefined phase change workflow
     */
    async executePhaseChangeWorkflow(
        itemId: string,
        newPhase: string,
        context: AgentContext
    ): Promise<WorkflowResult> {
        const workflow: AgentWorkflow = {
            name: 'phase-change',
            description: 'AI-powered phase change validation and prediction',
            steps: [
                {
                    id: 'predict',
                    agentName: 'PredictionAgent',
                    input: { itemId, targetType: 'item', currentPhase: newPhase },
                },
                {
                    id: 'recommend',
                    agentName: 'RecommendationAgent',
                    input: {
                        userId: context.userId,
                        contextType: 'item',
                        currentContext: { itemId, newPhase, predictions: '@predict' },
                    },
                    dependsOn: ['predict'],
                },
            ],
            stopOnError: false,
        };

        return this.executeWorkflow(workflow, context);
    }

    /**
     * Execute a predefined risk assessment workflow
     */
    async executeRiskAssessmentWorkflow(
        targetId: string,
        targetType: 'item' | 'phase' | 'workflow',
        context: AgentContext
    ): Promise<WorkflowResult> {
        const workflow: AgentWorkflow = {
            name: 'risk-assessment',
            description: 'Comprehensive risk assessment using multiple agents',
            steps: [
                {
                    id: 'predict',
                    agentName: 'PredictionAgent',
                    input: { targetId, targetType },
                },
                {
                    id: 'anomaly',
                    agentName: 'AnomalyDetectorAgent',
                    input: {
                        metricType: 'velocity',
                        currentMetrics: [],
                    },
                },
                {
                    id: 'copilot-summary',
                    agentName: 'CopilotAgent',
                    input: {
                        query: `Summarize the risk assessment for ${targetType} ${targetId}`,
                        selectedItems: [targetId],
                    },
                    dependsOn: ['predict', 'anomaly'],
                    optional: true,
                },
            ],
            stopOnError: false,
        };

        return this.executeWorkflow(workflow, context);
    }

    /**
     * Health check for all registered agents
     */
    async healthCheck(): Promise<Record<AgentName, 'healthy' | 'degraded' | 'unhealthy'>> {
        const results: Record<string, 'healthy' | 'degraded' | 'unhealthy'> = {};

        await Promise.all(
            Array.from(this.agents.entries()).map(async ([name, agent]) => {
                try {
                    const health = await agent.healthCheck();
                    results[name] = health.healthy ? 'healthy' : 'degraded';
                } catch {
                    results[name] = 'unhealthy';
                }
            })
        );

        return results as Record<AgentName, 'healthy' | 'degraded' | 'unhealthy'>;
    }
}

/**
 * Create a default orchestrator with common workflows
 */
export function createDefaultOrchestrator(): AgentOrchestrator {
    return new AgentOrchestrator();
}

File: products/yappc/core/yappc-domain-impl/src/main/java/com/ghatana/yappc/domain/agent/AgentOrchestrator.java
package com.ghatana.products.yappc.domain.agent;

import com.ghatana.platform.observability.MetricsCollector;
import io.activej.promise.Promise;
import io.activej.promise.Promises;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

/**
 * Agent Orchestrator for coordinating multi-agent workflows.
 * <p>
 * Manages the execution of complex workflows involving multiple AI agents
 * with dependency resolution, parallel execution, and error handling.
 *
 * @doc.type class
 * @doc.purpose Multi-agent workflow orchestration
 * @doc.layer product
 * @doc.pattern Orchestrator
 */
public class AgentOrchestrator {

    private static final Logger LOG = LoggerFactory.getLogger(AgentOrchestrator.class);

    private final Map<AgentName, AIAgent<?, ?>> agents;
    private final MetricsCollector metricsCollector;

    /**
     * Creates a new AgentOrchestrator.
     *
     * @param metricsCollector The metrics collector
     */
    public AgentOrchestrator(@NotNull MetricsCollector metricsCollector) {
        this.agents = new ConcurrentHashMap<>();
        this.metricsCollector = metricsCollector;
    }

    /**
     * Registers an agent with the orchestrator.
     *
     * @param agent The agent to register
     */
    public void registerAgent(@NotNull AIAgent<?, ?> agent) {
        AgentName name = agent.getMetadata().name();
        agents.put(name, agent);
        LOG.info("Registered agent: {}", name.getDisplayName());
    }

    /**
     * Unregisters an agent from the orchestrator.
     *
     * @param name The agent name to unregister
     */
    public void unregisterAgent(@NotNull AgentName name) {
        agents.remove(name);
        LOG.info("Unregistered agent: {}", name.getDisplayName());
    }

    /**
     * Gets an agent by name.
     *
     * @param name The agent name
     * @return The agent, or null if not found
     */
    @Nullable
    public AIAgent<?, ?> getAgent(@NotNull AgentName name) {
        return agents.get(name);
    }

    /**
     * Executes a workflow with multiple agent steps.
     *
     * @param workflow The workflow to execute
     * @param context  The execution context
     * @return Promise resolving to the workflow result
     */
    public Promise<WorkflowResult> executeWorkflow(
            @NotNull AgentWorkflow workflow,
            @NotNull AIAgentContext context
    ) {
        LOG.info("Starting workflow: {}", workflow.name());
        long startTime = System.currentTimeMillis();

        metricsCollector.incrementCounter("orchestrator.workflows.started", "workflow", workflow.name());

        Map<String, AgentResult<?>> results = new ConcurrentHashMap<>();

        // Build execution plan based on dependencies
        List<List<WorkflowStep>> executionPlan = buildExecutionPlan(workflow);

        return executeStages(executionPlan, results, context, workflow.stopOnError())
                .map(v -> {
                    long duration = System.currentTimeMillis() - startTime;

                    boolean allSuccess = results.values().stream().allMatch(AgentResult::success);

                    WorkflowResult result = new WorkflowResult(
                            allSuccess,
                            new ArrayList<>(results.values()),
                            aggregateMetrics(results),
                            duration
                    );

                    if (allSuccess) {
                        metricsCollector.incrementCounter("orchestrator.workflows.success", "workflow", workflow.name());
                    } else {
                        metricsCollector.incrementCounter("orchestrator.workflows.failed", "workflow", workflow.name());
                    }

                    metricsCollector.recordTimer("orchestrator.workflows.duration", duration, "workflow", workflow.name());

                    return result;
                })
                .mapException(error -> {
                    LOG.error("Workflow {} failed: {}", workflow.name(), error.getMessage());
                    metricsCollector.incrementCounter("orchestrator.workflows.error", "workflow", workflow.name());
                    return error;
                });
    }

    private List<List<WorkflowStep>> buildExecutionPlan(AgentWorkflow workflow) {
        // Topological sort based on dependencies
        Map<String, WorkflowStep> stepMap = new HashMap<>();
        Map<String, Set<String>> inDegree = new HashMap<>();

        for (WorkflowStep step : workflow.steps()) {
            stepMap.put(step.id(), step);
            inDegree.put(step.id(), new HashSet<>(step.dependsOn()));
        }

        List<List<WorkflowStep>> stages = new ArrayList<>();

        while (!stepMap.isEmpty()) {
            // Find all steps with no remaining dependencies
            List<WorkflowStep> currentStage = new ArrayList<>();

            for (Map.Entry<String, Set<String>> entry : new HashMap<>(inDegree).entrySet()) {
                if (entry.getValue().isEmpty()) {
                    currentStage.add(stepMap.get(entry.getKey()));
                }
            }

            if (currentStage.isEmpty() && !stepMap.isEmpty()) {
                throw new IllegalArgumentException("Circular dependency detected in workflow");
            }

            // Remove completed steps and update dependencies
            for (WorkflowStep step : currentStage) {
                stepMap.remove(step.id());
                inDegree.remove(step.id());

                // Update remaining steps that depended on this one
                for (Set<String> deps : inDegree.values()) {
                    deps.remove(step.id());
                }
            }

            if (!currentStage.isEmpty()) {
                stages.add(currentStage);
            }
        }

        return stages;
    }

    private Promise<Void> executeStages(
            List<List<WorkflowStep>> stages,
            Map<String, AgentResult<?>> results,
            AIAgentContext context,
            boolean stopOnError
    ) {
        Promise<Void> chain = Promise.complete();

        for (List<WorkflowStep> stage : stages) {
            chain = chain.then(v -> executeStage(stage, results, context, stopOnError));
        }

        return chain;
    }

    private Promise<Void> executeStage(
            List<WorkflowStep> stage,
            Map<String, AgentResult<?>> results,
            AIAgentContext context,
            boolean stopOnError
    ) {
        // Execute all steps in the stage in parallel
        List<Promise<Void>> stepPromises = new ArrayList<>();

        for (WorkflowStep step : stage) {
            Promise<Void> stepPromise = executeStep(step, results, context)
                    .then(result -> {
                        results.put(step.id(), result);

                        if (!result.success() && stopOnError) {
                            return Promise.ofException(new WorkflowStepException(step.id(), result.error()));
                        }

                        return Promise.complete();
                    });

            stepPromises.add(stepPromise);
        }

        return Promises.toList(stepPromises).toVoid();
    }

    @SuppressWarnings("unchecked")
    private <TInput, TOutput> Promise<AgentResult<?>> executeStep(
            WorkflowStep step,
            Map<String, AgentResult<?>> previousResults,
            AIAgentContext context
    ) {
        AIAgent<TInput, TOutput> agent = (AIAgent<TInput, TOutput>) agents.get(step.agentName());

        if (agent == null) {
            return Promise.ofException(new IllegalArgumentException("Agent not found: " + step.agentName()));
        }

        // Build input from dependencies
        TInput input = (TInput) buildInputFromDependencies(step, previousResults);

        return agent.execute(input, context)
                .map(result -> (AgentResult<?>) result);
    }

    private Object buildInputFromDependencies(
            WorkflowStep step,
            Map<String, AgentResult<?>> previousResults
    ) {
        Map<String, Object> resolvedInput = new HashMap<>(step.input());

        // Replace references to previous results
        for (Map.Entry<String, Object> entry : step.input().entrySet()) {
            Object value = entry.getValue();
            if (value instanceof String strValue && strValue.startsWith("@")) {
                String refId = strValue.substring(1);
                AgentResult<?> refResult = previousResults.get(refId);
                if (refResult != null && refResult.success()) {
                    resolvedInput.put(entry.getKey(), refResult.data());
                }
            }
        }

        return resolvedInput;
    }

    private WorkflowMetrics aggregateMetrics(Map<String, AgentResult<?>> results) {
        long totalLatency = 0;
        int totalTokens = 0;
        double totalCost = 0;

        for (AgentResult<?> result : results.values()) {
            totalLatency += result.metrics().latencyMs();
            if (result.metrics().tokensUsed() != null) {
                totalTokens += result.metrics().tokensUsed();
            }
            if (result.metrics().costUSD() != null) {
                totalCost += result.metrics().costUSD();
            }
        }

        return new WorkflowMetrics(
                totalLatency,
                totalTokens,
                totalCost,
                results.size(),
                (int) results.values().stream().filter(AgentResult::success).count()
        );
    }

    /**
     * Executes a phase change workflow involving multiple agents.
     */
    public Promise<WorkflowResult> executePhaseChangeWorkflow(
            String itemId,
            String newPhase,
            AIAgentContext context
    ) {
        AgentWorkflow workflow = new AgentWorkflow(
                "phase-change",
                "Phase Change Workflow",
                List.of(
                        new WorkflowStep(
                                "predict",
                                AgentName.PREDICTION_AGENT,
                                Map.of("itemId", itemId, "currentPhase", newPhase),
                                List.of()
                        ),
                        new WorkflowStep(
                                "analyze-risks",
                                AgentName.ANOMALY_DETECTOR_AGENT,
                                Map.of("itemId", itemId, "metricType", "VELOCITY"),
                                List.of()
                        ),
                        new WorkflowStep(
                                "recommend",
                                AgentName.RECOMMENDATION_AGENT,
                                Map.of(
                                        "itemId", itemId,
                                        "predictions", "@predict",
                                        "risks", "@analyze-risks"
                                ),
                                List.of("predict", "analyze-risks")
                        )
                ),
                false
        );

        return executeWorkflow(workflow, context);
    }

    /**
     * Executes a comprehensive item analysis workflow.
     */
    public Promise<WorkflowResult> executeItemAnalysisWorkflow(
            String itemId,
            AIAgentContext context
    ) {
        AgentWorkflow workflow = new AgentWorkflow(
                "item-analysis",
                "Comprehensive Item Analysis",
                List.of(
                        new WorkflowStep(
                                "predict",
                                AgentName.PREDICTION_AGENT,
                                Map.of("itemId", itemId, "horizonDays", 30),
                                List.of()
                        ),
                        new WorkflowStep(
                                "anomalies",
                                AgentName.ANOMALY_DETECTOR_AGENT,
                                Map.of("itemId", itemId, "metricType", "QUALITY_METRICS"),
                                List.of()
                        ),
                        new WorkflowStep(
                                "copilot-summary",
                                AgentName.COPILOT_AGENT,
                                Map.of(
                                        "query", "Summarize the analysis for item " + itemId,
                                        "predictions", "@predict",
                                        "anomalies", "@anomalies"
                                ),
                                List.of("predict", "anomalies")
                        )
                ),
                false
        );

        return executeWorkflow(workflow, context);
    }

    /**
     * Performs health checks on all registered agents.
     */
    public Promise<Map<AgentName, AgentHealth>> healthCheckAll() {
        Map<AgentName, Promise<AgentHealth>> healthPromises = new HashMap<>();

        for (Map.Entry<AgentName, AIAgent<?, ?>> entry : agents.entrySet()) {
            healthPromises.put(entry.getKey(), entry.getValue().healthCheck());
        }

        return Promises.toList(
                healthPromises.entrySet().stream()
                        .map(e -> e.getValue().map(h -> Map.entry(e.getKey(), h)))
                        .toList()
        ).map(entries -> {
            Map<AgentName, AgentHealth> result = new HashMap<>();
            for (Map.Entry<AgentName, AgentHealth> entry : entries) {
                result.put(entry.getKey(), entry.getValue());
            }
            return result;
        });
    }

    /**
     * Gets the status of all registered agents.
     */
    public List<AgentStatus> getAgentStatuses() {
        List<AgentStatus> statuses = new ArrayList<>();

        for (Map.Entry<AgentName, AIAgent<?, ?>> entry : agents.entrySet()) {
            AgentMetadata metadata = entry.getValue().getMetadata();
            statuses.add(new AgentStatus(
                    entry.getKey(),
                    metadata.version(),
                    true, // registered
                    metadata.latencySLA()
            ));
        }

        return statuses;
    }

    // Data structures

    /**
     * A workflow definition.
     */
    public record AgentWorkflow(
            @NotNull String name,
            @NotNull String description,
            @NotNull List<WorkflowStep> steps,
            boolean stopOnError
    ) {}

    /**
     * A single step in a workflow.
     */
    public record WorkflowStep(
            @NotNull String id,
            @NotNull AgentName agentName,
            @NotNull Map<String, Object> input,
            @NotNull List<String> dependsOn
    ) {}

    /**
     * Result of a workflow execution.
     */
    public record WorkflowResult(
            boolean success,
            @NotNull List<AgentResult<?>> results,
            @NotNull WorkflowMetrics metrics,
            long durationMs
    ) {}

    /**
     * Aggregated workflow metrics.
     */
    public record WorkflowMetrics(
            long totalLatencyMs,
            int totalTokensUsed,
            double totalCostUSD,
            int stepsExecuted,
            int stepsSucceeded
    ) {}

    /**
     * Agent status information.
     */
    public record AgentStatus(
            @NotNull AgentName name,
            @NotNull String version,
            boolean registered,
            long latencySLA
    ) {}

    /**
     * Exception for workflow step failures.
     */
    public static class WorkflowStepException extends RuntimeException {
        private final String stepId;
        private final AgentResult.AgentError error;

        public WorkflowStepException(String stepId, AgentResult.AgentError error) {
            super("Workflow step " + stepId + " failed: " + (error != null ? error.message() : "Unknown error"));
            this.stepId = stepId;
            this.error = error;
        }

        public String getStepId() {
            return stepId;
        }

        public AgentResult.AgentError getError() {
            return error;
        }
    }
}
